src/core/analysis-context.ts

```diff
@@ -498,4 +489,28 @@ export class AnalysisContext {
     return this.resolver.canResolve(url) ? this.resolver.resolve(url) :
                                            (url as any as ResolvedUrl);
   }
+
+  /**
+   * A warning for a weird situation that should never happen.
+   *
+   * Before calling getDocument(), which is synchronous, a caller must first
+   * have finished loading and scanning, as those phases are asynchronous.
+   *
+   * So we need to construct a warning, but we don't have a parsed document,   *
+   * so we construct this weird fake one.
```

Comment:
todo and/or issue if we want to have document-less warnings?

Reply:
I think this is working as intended for now. Better to have this awkward function than to make all uses of Warning more awkward or slow.